### DISCUSSION TITLE
Make pre-commit a default and not an extension

### DISCUSSION THREAD

**Discussion by CarliJoy** [pyscaffold/pyscaffold#356] [Ideas]
It was [suggested](https://github.com/pyscaffold/pyscaffold/discussions/336#discussioncomment-243888) to make pre-commit not an extension but include it by default in the creation of a project.

It would not interfere very with existing projects but would allow to [use pre-commit in the CI](https://pre-commit.com/#usage-in-continuous-integration).
Doing so the users would get a more streamlined experience as a big chuncks of the test that run on the CI will run locally already reducing waiting for failing CI services.

I don't see a disadvantage add .pre-commit as default.
Even so pre-commit can be a bit of a pain in the ass if you have to use it behind a restrictive corporate firewall, it is still optional as you simply could not run `pre-commit init` for the given repro.

**Comment by abravalheri** [comment]
Thank you very much @CarliJoy for separating the topics.

Somethings to have in mind for the discussion:

- v4 will automatically attempt to run `pre-commit install` at the end of `putup` (best effort), so it is no longer that easy to _decide to not run `pre-commit init` for the given repo_.
- I imagine that the new `${XDG_CONFIG_HOME:-"$HOME/.config"}/pyscaffold/setup.cfg` file and `--save-config` option will make it easier to deal with "favourite/recurring configs" such as `pre-commit` blurring the line between default and opt-in features (at least in terms of day to day effort to use them), making the life of people that have to run `putup --update` just because they forgot an extension easier.

Considering the restrictive corporate firewall problem, leaving `pre-commit` optional might work just well for the majority of the use cases...

**Comment by CarliJoy** [comment]
Maybe it would easy to include a switch "--no-pre-commit-init" the solution?
That v4 would not run "pre-commit install" at the end (opt out).
I really see the benefit of using pre-commit as default, as other extension could rely on it, reducing the possible branches.

And besides the cooperate firewall problem I don't see any reason why not to use it.

In the end it is a decision what to we consider as standard and what not.

BTW: v4 sounds better from time to time I duck into the details. Thanks @abravalheri putting that much work into it! Would be happy to contribute also besides putting my discussion input into it.
Will have a look on the CI stuff now...

**Comment by abravalheri** [comment]
Hi @CarliJoy thank you very much for the enthusiasm, discussions and help!
It is really nice to have more people involved and give us more stamina to work on the changes 😄 

I think for v4, we would need to prioritize a bit the work (my guess is that around March or April we are going to see new big changes in setuptools so it would be cool if we have v4 battle tested by that date to start working on v5, moreover its been a good amount of months v4 is on the making, I cannot wait to see it out 😝). 

With that in mind I updated the [v4 milestone](https://github.com/pyscaffold/pyscaffold/milestone/18) with the issues I think would be the best to work with  (please feel free to have a look and if you feel like contribute, you are super welcome).

I have a feeling that the new config file will make it much easier to work with our favourite extensions such as `pre-commit` and that is the main reason why I am delaying this decision a little bit. As the wise Florian always say, "let's release it and wait to see the feedback".

**Comment by julie777** [comment]
I think that precommit is a best practice and in line with what PyScaffold is trying to provide for users.

If we generate a project with a precommit hook and the user doesn't want it why can't they just delete the precommit file and because it doesn't exist pyscaffold won't try to use it.

I'm not sure I understand the corporate firewall issue as precommit should be running before I commit locally. It should be all on my machine. Continuous Integration tests, or processing as part of a pull request are a different matter.

Am I just confused?

**Reply by CarliJoy** [reply]
Yep. The problem is that in you can't access external repos if you are behind a corporate firewall. Therefore pre-commit will always fill (locally on your machine) if you try to commit or push.

But this is an issue that coporate users of pyscaffold should fix with an extension, as did I.

Only option for pyscaffold would be to add an option that allows to define internal mirrors for all pre-commit repos used.
